Training24/FSMTest.c

```diff
@@ -0,0 +1,92 @@
+// FSMTest.c
+// Program on A6b branch.
+// ------------------------------------------------------------------------------------
+#define _CRT_SECURE_NO_WARNINGS 1
+
+#include <malloc.h>
+#include <stdio.h>
+#include <string.h>
+#include <windows.h>
+
+/// <summary>To run the program with the given input and compare output with expected.</summary>
+int RunTestProgram (const char* exeFilePathAndName, const char* inputFilePathAndName, const char* outputFilePathAndName);
+
+/// <summary>To compare two files.</summary>
+int CompareFiles (const char* file1, const char* file2);
+
+int RunTestProgram (const char* exeFilePathAndName, const char* inputFilePathAndName, const char* outputFilePathAndName) {
+   char* cmdline = malloc (strlen (exeFilePathAndName) + strlen (inputFilePathAndName) + strlen (outputFilePathAndName) + 3);
+   if (cmdline == NULL) {
+      printf ("Unable to allocate memory\n");
+      return 1;
+   }
+   sprintf (cmdline, "%s %s %s", exeFilePathAndName, inputFilePathAndName, outputFilePathAndName);
+   STARTUPINFOA si = { 0 };
+   si.cb = sizeof (si);
+   PROCESS_INFORMATION pi = { 0 };
+   if (!CreateProcessA (NULL, cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi)) {
+      printf ("Failed to start process. Error: %lu\n", GetLastError ());
+      free (cmdline);
+      return 1;
+   }
+   WaitForSingleObject (pi.hProcess, INFINITE);
+   CloseHandle (pi.hProcess);
+   CloseHandle (pi.hThread);
+   free (cmdline);
+   return 0;
+}
+
+int CompareFiles (const char* file1, const char* file2) {
+   FILE* f1 = fopen (file1, "r"), * f2 = fopen (file2, "r");
+   if (!f1 || !f2) {
+      printf ("Error opening files.\n");
+      return 1;
+   }
+   int pos = 0, ch1, ch2;
+   while ((ch1 = fgetc (f1)) != EOF && (ch2 = fgetc (f2)) != EOF) {
+      if (ch1 != ch2) {
+         printf ("Error at bit no. %d, Expected %c, Actual %c\n", pos, ch1, ch2);
+         fclose (f1);
+         fclose (f2);
+         return 1;
+      }
+      pos++;
+   }
+   fclose (f1);
+   fclose (f2);
+   return 0;
+}
+
+int main (int argc, char** argv) {
+   if (argc != 2) {
+      printf ("Usage: %s <FSM executable name>\n", argv[0]);
+      return -1;
+   }
+   const char* folderPath = "TData/",
+      * inputFiles[] = { "TData/Test1in.txt", "TData/Test2in.txt", "TData/Test3in.txt", "TData/Test4in.txt",
+         "TData/Test5in.txt", "TData/Test6in.txt", "TData/Test7in.txt", "TData/Test8in.txt" },
+      * expectedFiles[] = { "TData/Test1ref.txt", "TData/Test2ref.txt", "TData/Test3ref.txt", "TData/Test4ref.txt",
+         "TData/Test5ref.txt", "TData/Test6ref.txt", "TData/Test7ref.txt", "TData/Test8ref.txt" };
+   int numTests = sizeof (inputFiles) / sizeof (inputFiles[0]);
+   for (int i = 0; i < numTests; i++) {
+      const char* inputFile = inputFiles[i], * expectedFile = expectedFiles[i];
+      // Generate a temporary output file name for each test
+      char outputFile[50];
+      sprintf (outputFile, "temp%d.txt", i + 1);
```

Comment:
You don't need to create a new output file every time. It can be a temp file that is overwritten for each test since this is not something we need to store.